Source/Core/S2Cell.js

```diff
@@ -0,0 +1,310 @@
+/* eslint-disable no-undef */
+import Check from "./Check.js";
+import defined from "./defined.js";
+import DeveloperError from "./DeveloperError.js";
+import FeatureDetection from "./FeatureDetection.js";
+import RuntimeError from "./RuntimeError.js";
+
+// The maximum level supported within an S2 cell ID. Each level is represented by two bits in the
+// final cell ID
+var S2MaxLevel = 30;
+
+// The number of bits in a S2 cell ID used for specifying the base face
+var S2FaceBits = 3;
+
+// The number of bits in a S2 cell ID used for specifying the position along the Hilbert curve
+var S2PositionBits = 2 * S2MaxLevel + 1;
+
+// The number of bits per I and J in the lookup tables
+var S2LookupBits = 4;
+
+// Lookup table for mapping 10 bits of IJ + orientation to 10 bits of Hilbert curve position +
+// orientation.
+var S2LookupPositions;
+
+// Lookup table for mapping 10 bits of IJ + orientation to 10 bits of Hilbert curve position +
+// orientation.
+var S2LookupIJ;
+
+// Lookup table of two bits of IJ from two bits of curve position, based also on the current curve
+// orientation from the swap and invert bits
+var S2PosToIJ = [
+  [0, 1, 3, 2], // 0: Normal order, no swap or invert
+  [0, 2, 3, 1], // 1: Swap bit set, swap I and J bits
+  [3, 2, 0, 1], // 2: Invert bit set, invert bits
+  [3, 1, 0, 2], // 3: Swap and invert bits set
+];
+
+// Mask that specifies the swap orientation bit for the Hilbert curve
+var S2SwapMask = 1;
+
+// Mask that specifies the invert orientation bit for the Hilbert curve
+var S2InvertMask = 2;
+
+// Lookup for the orientation update mask of one of the four sub-cells within a higher level cell.
+// This mask is XOR'ed with the current orientation to get the sub-cell orientation.
+var S2PosToOrientationMask = [S2SwapMask, 0, 0, S2SwapMask | S2InvertMask];
+
+/**
+ * Represents a cell in the S2 geometry library.
+ *
+ * @alias S2Cell
+ * @constructor
+ *
+ * @param {String} [token] The hexadecimal representation of an S2CellId. 0 must be represented using "X".
```

Comment:
Remember to add `@private` if we don't plan to make this available in the documentation.